lib/jyson.js

```diff
@@ -1,19 +1,13 @@
 const assert = require('assert');
 const _ = require('lodash');
 
-const getKey = (object, key, undefinedValue) => {
-  const arraySplitKeys = key.split('.$');
-  const value = _.get(object, arraySplitKeys.shift(), undefinedValue);
+const getKey = (object, key, undefinedValue, opts) => {
+  if (key.indexOf('$')!== -1) {
+    const arrayKey = key.split('.$').shift();
 
-  // We have found our key
-  if (arraySplitKeys.length === 0) {
-    return value;
+    assert(_.get(object, arrayKey).length > opts.arrayIndex['$'], 'We hit the end of the array');
```

Comment:
Not sure how I feel about this

Reply:
Looks like this function returns a value and not a key? If that's the case you may want to rename it to `getValue` or something similar.

Reply:
How is `opts.arrayIndex` used?  Could the `arrayIndex` value be stored outside of `opts`? It seems that `opts` stands for "options", but it's being used to track current index?

Reply:
> Looks like this function returns a value and not a key? If that's the case you may want to rename it to `getValue` or something similar.

If you are referring to `getKey` it returns a value

Reply:
> How is `opts.arrayIndex` used?  Could the `arrayIndex` value be stored outside of `opts`? It seems that `opts` stands for "options", but it's being used to track current index?

@etorrejon great point. `opts` is passed to functions https://github.com/hubba/jyson/blob/master/spec/lib/jyson/jyson.function.spec.js#L52 as they will need to to get the correct data.

Do you think we should rename opts? Should we add an additional argument to functions?

Reply:
@misshula @etorrejon My original comment however was about the `assert` Im not sure that relying on an error to exit out of a loop is a good pattern. Thoughts?

Reply:
@earobinson 
- yup, I was referring to `getKey`. `getValue` would be more intuitive.
- my question about the `arrayKey` was whether it actually needs to be part of `opts`. It seems as though it doesn't, since it's used to track an array index and not a configuration option?

Reply:
also... re: your original question, I think it's fine to error out in that case

Reply:
... if jyson were attempting to read an index that was out of bounds, we'd likely want to know.

Reply:
> ... if jyson were attempting to read an index that was out of bounds, we'd likely want to know.

@etorrejon 

Issue is we catch the error here https://github.com/hubba/jyson/pull/11/files#diff-939716d327873b484c0914e8a74fc40bR52